File: solutions/0477. Total Hamming Distance/0477.java
class Solution {
  public int totalHammingDistance(int[] nums) {
    int ans = 0;
    int mask = 1;

    for (int i = 0; i < 30; ++i) {
      final int onesCount = getCount(nums, mask);
      final int zerosCount = nums.length - onesCount;
      ans += onesCount * zerosCount;
      mask <<= 1;
    }

    return ans;
  }

  private int getCount(int[] nums, int mask) {
    int count = 0;
    for (final int num : nums)
      if ((num & mask) > 0)
        ++count;
    return count;
  }
}

File: solutions/0477. Total Hamming Distance/0477.cpp
class Solution {
 public:
  int totalHammingDistance(vector<int>& nums) {
    constexpr int kMaxMask = 1 << 30;
    int ans = 0;

    for (int mask = 1; mask < kMaxMask; mask <<= 1) {
      const int ones =
          ranges::count_if(nums, [mask](int num) { return num & mask; });
      const int zeros = nums.size() - ones;
      ans += ones * zeros;
    }

    return ans;
  }
};
